### PR TITLE
TLS server support

### PR DESCRIPTION
**Pull request by Geal**
This allows TLS for the GraphQL endpoint and any other endpoint using the same listen address. For now, the configuration is limited to one server certificate, TLS 1.2 and 1.3 (rustls does not support lower versions), safe default ciphers and no client authentication.

Fixes https://github.com/apollographql/router/issues/2615

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [x] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

### CODE REVIEW

File: apollo-router/src/configuration/mod.rs [hunk 1]
```diff
@@ -622,9 +632,140 @@ pub(crate) struct RedisCache {
...
+    certs(&mut BufReader::new(data.as_bytes()))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid cert"))
+        .map(|mut certs| certs.drain(..).map(Certificate).collect())
+}
+
+//FIXME: handle ECDSA keys too
+fn load_keys(data: &str) -> io::Result<PrivateKey> {
+    /*rsa_private_keys(&mut BufReader::new(data.as_bytes()))
+    .map_err(|_| serde::de::Error::custom(io::Error::new(io::ErrorKind::InvalidInput, "invalid key")))
+    .map(|mut keys| keys.drain(..).map(PrivateKey).collect())*/
```
Comment (bnjjj) [inline apollo-router/src/configuration/mod.rs:725]: Should we delete this ?

File: apollo-router/src/configuration/mod.rs [hunk 2]
```diff
@@ -622,9 +632,140 @@ pub(crate) struct RedisCache {
...
+
+    let mut keys = rsa_private_keys(&mut BufReader::new(data.as_bytes()))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid key"))?;
+
+    if keys.len() > 1 {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "expected exactly one private key",
```
Comment (bnjjj) [inline apollo-router/src/configuration/mod.rs:733]:
```suggestion
            "expected exactly one private rsa key",
```

File: apollo-router/src/configuration/mod.rs [hunk 3]
```diff
@@ -622,9 +632,140 @@ pub(crate) struct RedisCache {
...
+        return Ok(PrivateKey(key));
+    }
+
+    let mut keys = rustls_pemfile::pkcs8_private_keys(&mut BufReader::new(data.as_bytes()))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid key"))?;
+
+    if keys.len() > 1 {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "expected exactly one private key",
```
Comment (bnjjj) [inline apollo-router/src/configuration/mod.rs:745]:
```suggestion
            "expected exactly one private pkcs8 key",
```

File: apollo-router/src/configuration/mod.rs [hunk 4]
```diff
@@ -622,9 +632,140 @@ pub(crate) struct RedisCache {
...
+        return Ok(PrivateKey(key));
+    }
+
+    let mut keys = rustls_pemfile::ec_private_keys(&mut BufReader::new(data.as_bytes()))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid key"))?;
+
+    if keys.len() > 1 {
+        Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "expected exactly one private key",
```
Comment (bnjjj) [inline apollo-router/src/configuration/mod.rs:757]:
```suggestion
            "expected exactly one private ec key",
```

File: apollo-router/src/configuration/mod.rs [hunk 5]
```diff
@@ -622,9 +632,140 @@ pub(crate) struct RedisCache {
...
+}
+
+//FIXME: handle ECDSA keys too
+fn load_keys(data: &str) -> io::Result<PrivateKey> {
+    /*rsa_private_keys(&mut BufReader::new(data.as_bytes()))
+    .map_err(|_| serde::de::Error::custom(io::Error::new(io::ErrorKind::InvalidInput, "invalid key")))
+    .map(|mut keys| keys.drain(..).map(PrivateKey).collect())*/
+
+    let mut keys = rsa_private_keys(&mut BufReader::new(data.as_bytes()))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid key"))?;
```
Comment (bnjjj) [inline apollo-router/src/configuration/mod.rs:728]:
```suggestion
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid rsa key"))?;
```

File: apollo-router/src/configuration/mod.rs [hunk 6]
```diff
@@ -622,9 +632,140 @@ pub(crate) struct RedisCache {
...
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "expected exactly one private key",
+        ));
+    } else if let Some(key) = keys.pop() {
+        return Ok(PrivateKey(key));
+    }
+
+    let mut keys = rustls_pemfile::pkcs8_private_keys(&mut BufReader::new(data.as_bytes()))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid key"))?;
```
Comment (bnjjj) [inline apollo-router/src/configuration/mod.rs:740]:
```suggestion
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid pkcs8 key"))?;
```

File: apollo-router/src/configuration/mod.rs [hunk 7]
```diff
@@ -622,9 +632,140 @@ pub(crate) struct RedisCache {
...
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "expected exactly one private key",
+        ));
+    } else if let Some(key) = keys.pop() {
+        return Ok(PrivateKey(key));
+    }
+
+    let mut keys = rustls_pemfile::ec_private_keys(&mut BufReader::new(data.as_bytes()))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid key"))?;
```
Comment (bnjjj) [inline apollo-router/src/configuration/mod.rs:752]:
```suggestion
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid ec key"))?;
```

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/configuration/mod.rs [hunk 8]
```diff
@@ -623,9 +635,130 @@ pub(crate) struct RedisCache {
...
+    load_keys(&data).map_err(serde::de::Error::custom)
+}
+
+fn load_certs(data: &str) -> io::Result<Vec<Certificate>> {
+    certs(&mut BufReader::new(data.as_bytes()))
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid cert"))
+        .map(|mut certs| certs.drain(..).map(Certificate).collect())
+}
+
+//FIXME: handle ECDSA keys too
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/configuration/mod.rs:724]: that's a comment I forgot to remove
```suggestion
```

**Pull request merged**